File: 144-binary-tree-preorder-traversal/binary-tree-preorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> ans=new ArrayList<>();
        Stack<TreeNode> q=new Stack<>();
        TreeNode ptr=root;
        if(root==null)
           return ans;
        if(root.right!=null)
           q.push(root);
        ans.add(root.val);
        while(ptr!=null ||!q.isEmpty()){
            while(ptr.left!=null){
                ptr=ptr.left;
                ans.add(ptr.val);
                if(ptr.right!=null){
                    q.push(ptr);
                }
            }
            if(!q.isEmpty()){
               ptr=q.pop();
               ptr=ptr.right;
               if(ptr.right!=null){
                   q.push(ptr);
               }
               ans.add(ptr.val); 
            }
            else break;
        }
        return ans;
    }
}

File: 144-binary-tree-preorder-traversal/binary-tree-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """

        stack = []
        ans = []
        ptr = root

        while True:
            while ptr != None:
                ans.append(ptr.val)
                if ptr.right != None:
                    stack.append(ptr.right)
                ptr = ptr.left
            
            if len(stack) != 0:
                ptr = stack[-1]
                stack.pop(-1)
            else:
                break
            
        return ans

        